docs/design/endpoints.md

```diff
@@ -271,16 +471,17 @@ public class DefaultSmithyServiceEndpointProvider: SmithyServiceEndpointProvider
     }
 }
 
-public class GetResourceResolveEndpointMiddleware(
+// package *.services.smithyservice.internal
+
+internal class GetResourceResolveEndpointMiddleware(
     private val config: SmithyService.Config,
     private val request: GetResourceRequest,
```

Comment:
Something we need to be careful of here now that I'm thinking about it is that the [initialize](https://github.com/awslabs/smithy-kotlin/blob/cfa0fd3a30b4c50b75485786f043d4e2ad803f55/runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/operation/SdkOperationExecution.kt#L44) middleware phase is allowed to act on the operation input (possibly modifying it pre-serialization). We wouldn't be able to just capture the operation input at the time middleware is registered as it could change.

e.g.

```kotlin
suspend fun fooOperation(input: FooInput): FooOutput {
     val op = SdkHttpOperation<FooInput, FooOutput>.build { ... }
     
     op.install(Middleware1())
     ...
     op.install(MiddlewareN())

      // example inline middleware to demonstrate this case, same thing applies to non-inlined middleware
      op.execution.initialize.intercept { req, next ->
          val modifiedFooInput = req.subject.copy { foobar = "baz" }
          val modifiedReq = req.copy(subject = modifiedFooInput)
          next.call(modifiedReq)
      }

     // captures the original input not the one modified by the middleware that will be serialized
     op.install(FooServiceEndpointMiddleware(config, input))

}
```

Now one way around this would be to have the service endpoint middleware implement both `initialize` and `modify` phases. If it's the last `initialize` middleware to run it can capture the finalized operation input that is used during the `modify` phase.